### DISCUSSION TITLE
Audioplayer with VolumePrint

### DISCUSSION THREAD

**Discussion by DrNeuroSurg** [pschatzmann/arduino-audio-tools#1518] [Q&A] [answered]
I'm using AudioPlayer, this works like a charm !!

So this is my chain:
...
ICYStream urlStream(wifi, password);
AudioSourceURL source(urlStream, urls, "audio/mp3");
I2SStream i2s;
MP3DecoderHelix decoder;
AudioPlayer player(source, i2s, decoder);
....

Now I would get the volume ... 
How is it possible to add an additional MultiOutput to get the volume via VolumeOutput ? 
Have I to subclass the AudioPlayer ? Or is there an easier way ?

 urlstream - > decoder -> multi -> volume -> i2s
                                                     |
                                                     -> VolumeOutput

**Comment by pschatzmann** [comment, marked as answer]
Just define the MultiOutput as output instead of i2s when you create the player

**Reply by DrNeuroSurg** [reply]
Thank you ! 
Is this the right way ?

VolumePrint vp;
MultiOutput multi(i2s, vp);
AudioPlayer player(source, multi, decoder);

**Comment by pschatzmann** [comment]
Yes that should work: if you have more then 2 outputs, you would need to add the output with the add method in the setup();

**Reply by DrNeuroSurg** [reply]
Thanx ... btw. where can I get the AudioInfo for "mlti.begin()"  ?? Source ?

**Reply by pschatzmann** [reply]
In this case you would not need to care and just leave it to the default by calling mlti.begin(). 

The decoder will call setAudioInfo() on the MultiOutput using the information from the encoded file and MultiOutput will then call setAudioInfo() on it's components.

**Reply by DrNeuroSurg** [reply]
Thank you again - but: MultiOutput.begin(); doesn't work, because 
bool begin(AudioInfo info) is defined ...

**Reply by pschatzmann** [reply]
I don't understand: what is your problem with calling begin() ? It is [available in the documentation](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_multi_output.html)...

**Reply by DrNeuroSurg** [reply]
Hmm. begin() is overridden in MultiOutput ..
I did this: multi.begin(player.audioInfo());

**Reply by DrNeuroSurg** [reply]
Another observation (if I'm right): 

If I do this
AudioPlayer player(source, multi, decoder);
then the MultiOutput is inserted AFTER the (internal) VolumeControl in AudioPlayer ...

So I have
AudioSource -> Stream-copy -> EncodedAudioStream -> VolumeStream -> FadeStream ->  MultiOutput -> Print

What I want to have is:

AudioSource -> Stream-copy -> EncodedAudioStream -> MultiOutput -> VolumeStream -> FadeStream -> Print
